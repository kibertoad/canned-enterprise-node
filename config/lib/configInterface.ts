export enum Brand {
  MainBrand = 'mainbrand',
  OtherBrand = 'otherbrand'
}

export enum Country {
  US = 'US',
  RUSSIA = 'RU'
}

export interface Config {
  aws: {
    accessKeyId: string
    secretAccessKey: string
    region: string
  }
  someExternalService: {
    baseUrl: string
    auth: {
      username: string
      password: string
    }
    callTimeout: number
  }
}

export interface ConfigDefinition {
  aws: {
    accessKeyId: StringParam
    secretAccessKey: StringParam
    region: StringParam
  }
  someExternalService: {
    baseUrl: StringParam
    auth: {
      username: StringParam
      password: StringParam
    }
    callTimeout: NumberParam
  }
}

export type ConfigDefinitionMap = DeepPartial<
  { [key in Environment | 'default']: ConfigDefinition }
>

type StringParam =
  | string
  | { [key in Brand]: string | { [key in Country]: string } }
  | { [key in Country]: string }
type BooleanParam = boolean | { [key in Brand]: boolean | { [key in Country]: boolean } | string[] }
type NumberParam = number | { [key in Brand]: number | { [key in Country]: number } | number[] }

export enum Environment {
  production = 'production',
  staging = 'staging',
  test = 'test',
  development = 'development'
}

export interface FeatureToggles {
  moduleOne: {
    SOME_FEATURE: boolean
  }
  moduleTwo: {
    ANOTHER_FEATURE: boolean
  }
  GLOBAL_FEATURE: boolean
}

export interface FeatureToggleDefinition {
  moduleOne: {
    SOME_FEATURE: BooleanParam | string[]
  }
  moduleTwo: {
    ANOTHER_FEATURE: BooleanParam | string[]
  }
  GLOBAL_FEATURE: BooleanParam | string[]
}

export type FeatureToggleEnvironmentMap = DeepPartial<
  { [key in Environment]: FeatureToggleDefinition }
>

export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Array<infer U>
    ? Array<DeepPartial<U>>
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : DeepPartial<T[P]>
}
