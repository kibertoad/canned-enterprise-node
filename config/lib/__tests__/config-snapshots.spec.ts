import { Brand, Country, Environment } from '../configInterface'

describe('config', () => {
  let initialEnv = process.env.APP_ENV

  beforeAll(() => {
    initialEnv = process.env.APP_ENV
  })

  afterAll(() => {
    process.env.APP_ENV = initialEnv
  })

  it('values are correct', async () => {
    for await (const environment of Object.values(Environment)) {
      process.env.APP_ENV = environment
      jest.resetModules()
      const configService = await import('../configService')

      Object.values(Brand).forEach(brand => {
        Object.values(Country).forEach(country => {
          expect(configService.getConfig(country, brand)).toMatchSnapshot(
            `Config for ${environment}: ${brand} ${country}`
          )
        })
      })
    }
  })

  it('feature toggle values are correct', async () => {
    for await (const environment of Object.values(Environment)) {
      process.env.APP_ENV = environment
      jest.resetModules()
      const configService = await import('../configService')

      Object.values(Brand).forEach(brand => {
        Object.values(Country).forEach(country => {
          expect(configService.getFeatures(country, brand)).toMatchSnapshot(
            `Features for ${environment}: ${brand} ${country}`
          )
        })
      })
    }
  })
})
