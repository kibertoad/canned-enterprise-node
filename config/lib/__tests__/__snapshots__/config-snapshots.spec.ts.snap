// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config feature toggle values are correct: Features for development: mainbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for development: mainbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for development: otherbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config feature toggle values are correct: Features for development: otherbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config feature toggle values are correct: Features for production: mainbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for production: mainbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for production: otherbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config feature toggle values are correct: Features for production: otherbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config feature toggle values are correct: Features for staging: mainbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": true,
  "moduleOne": Object {
    "SOME_FEATURE": true,
  },
  "moduleTwo": Object {
    "ANOTHER_FEATURE": true,
  },
}
`;

exports[`config feature toggle values are correct: Features for staging: mainbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": true,
  "moduleOne": Object {
    "SOME_FEATURE": true,
  },
  "moduleTwo": Object {
    "ANOTHER_FEATURE": true,
  },
}
`;

exports[`config feature toggle values are correct: Features for staging: otherbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": true,
  "moduleOne": Object {
    "SOME_FEATURE": true,
  },
  "moduleTwo": Object {
    "ANOTHER_FEATURE": true,
  },
}
`;

exports[`config feature toggle values are correct: Features for staging: otherbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": true,
  "moduleOne": Object {
    "SOME_FEATURE": true,
  },
  "moduleTwo": Object {
    "ANOTHER_FEATURE": true,
  },
}
`;

exports[`config feature toggle values are correct: Features for test: mainbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for test: mainbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": false,
}
`;

exports[`config feature toggle values are correct: Features for test: otherbrand RU 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config feature toggle values are correct: Features for test: otherbrand US 1`] = `
Object {
  "GLOBAL_FEATURE": true,
}
`;

exports[`config values are correct: Config for development: mainbrand RU 1`] = `Object {}`;

exports[`config values are correct: Config for development: mainbrand US 1`] = `Object {}`;

exports[`config values are correct: Config for development: otherbrand RU 1`] = `Object {}`;

exports[`config values are correct: Config for development: otherbrand US 1`] = `Object {}`;

exports[`config values are correct: Config for production: mainbrand RU 1`] = `
Object {
  "aws": Object {
    "region": "eu-north-1",
  },
}
`;

exports[`config values are correct: Config for production: mainbrand US 1`] = `
Object {
  "aws": Object {
    "region": "eu-west-1",
  },
}
`;

exports[`config values are correct: Config for production: otherbrand RU 1`] = `
Object {
  "aws": Object {
    "region": "eu-west-1",
  },
}
`;

exports[`config values are correct: Config for production: otherbrand US 1`] = `
Object {
  "aws": Object {
    "region": "eu-north-1",
  },
}
`;

exports[`config values are correct: Config for staging: mainbrand RU 1`] = `
Object {
  "aws": Object {
    "region": "eu-north-1",
  },
}
`;

exports[`config values are correct: Config for staging: mainbrand US 1`] = `
Object {
  "aws": Object {
    "region": "eu-west-1",
  },
}
`;

exports[`config values are correct: Config for staging: otherbrand RU 1`] = `
Object {
  "aws": Object {
    "region": "eu-north-1",
  },
}
`;

exports[`config values are correct: Config for staging: otherbrand US 1`] = `
Object {
  "aws": Object {
    "region": "eu-west-1",
  },
}
`;

exports[`config values are correct: Config for test: mainbrand RU 1`] = `Object {}`;

exports[`config values are correct: Config for test: mainbrand US 1`] = `Object {}`;

exports[`config values are correct: Config for test: otherbrand RU 1`] = `Object {}`;

exports[`config values are correct: Config for test: otherbrand US 1`] = `Object {}`;
